docs: move docs about receipts from confluence (#7947)
near/nearcore: docs/architecture/how/tx_receipts.md

# Transaction, Receipts adn Chunk Surprises

We finished the previous article ([Transaction routing](./tx_routing.md)) on the
part, where transaction was successfully added to the soon-to-be block
producer’s mempool.

In this article, we’ll cover what happens next: how it is changed into a receipt
and executed, potentially creating even more receipts in the process.

First, let’s look at the ‘high level view’:

![image](https://user-images.githubusercontent.com/1711539/198282472-3883dcc1-77ca-452c-b21e-0a7af1435ede.png)

## Transaction vs receipt

> As you can see form the image above:

Reviewer: ```suggestion
As you can see from the image above:
```

**Transactions** are ‘external’ communication - they are coming from the
outside.

> **Receipts** are used for  ‘internal’ communication (cross shard, cross

Reviewer: ```suggestion
**Receipts** are used for ‘internal’ communication (cross shard, cross
```

contract) - they are created by the block/chunk producers.

## Life of a Transaction

> If we ‘zoom-in', the chunk producer work looks like this:

Reviewer: ```suggestion
If we ‘zoom-in', the chunk producer's work looks like this:
```

![image](https://user-images.githubusercontent.com/1711539/198282518-cdeb375e-8f1c-4634-842c-6490020ad9c0.png)

### Step 1: Process Transaction into receipt

> Once chunk producer is ready to produce the chunk, it will fetch the

Reviewer: ```suggestion
Once a chunk producer is ready to produce a chunk, it will fetch the
```

transactions from its mempool, check that they are valid, and if so, prepare to

> change them into receipts.

Reviewer: ```suggestion
process them into receipts.
```

> **Note:** There are additional restrictions (making sure that we take them in
> the right order, that we don’t take too many etc) - that you can see in

Reviewer: ```suggestion
**Note:** There are additional restrictions (e.g. making sure that we take them in
the right order, that we don’t take too many, etc.) - that you can see in
```

> nomicon’s transaction page.

Reviewer: Should we add a link here?

> nomicon’s transaction page.

Author: ```suggestion
nomicon’s [transaction page](https://nomicon.io/ChainSpec/Transactions).
```

> You can see this part very clearly in explorer:

Reviewer: ```suggestion
You can see this part in explorer:
```

### Step 2: Sending receipt to the proper destination

Once we have a receipt, we have to send it to the proper destination - by adding
it to the ‘outgoing_receipt’ list, which will be forwarded to the chunk
producers from the next block.

**Note:** There is a special case here - if the sender of the receipt is the
same as the receiver, then the receipt will be added to the ‘local_receipts'
queue and executed in the same block.

### Step 3: When incoming receipt arrives

(Note: this happens in the ‘next’ block)

When chunk producer receives the incoming receipt, it will try to execute its
actions (creating accounts, executing function calls etc).

Such actions might generate additional receipts (for example a contract might
want to call other contracts). All these outputs are added to the outgoing
receipt queue to be executed in the next block.

If there is incoming receipt queue is too large to execute in the current chunk,
the producer will put the remaining receipts onto the ‘delayed’ queue.

### Step 4: Profit

When all the ‘dependant’ receipts are executed for a given transaction, we can
consider the transaction to be successful.

### [Advanced] But reality is more complex

> **Caution:** In the section below, some things are simplified and are not 100%
> match to how the current code works.

Reviewer: ```suggestion
**Caution:** In the section below, some things are simplified and do not match exactly 
to how the current code works.
```

Let’s quickly also check what’s actually inside a Chunk:

```rust
pub struct ShardChunkV2 {
    pub chunk_hash: ChunkHash,
    pub header: ShardChunkHeader,
    pub transactions: Vec<SignedTransaction>,
    pub receipts: Vec<Receipt>, // outgoing receipts from 'previous' block
}
```

> Yes, it is a little bit confusing, that receipts here are NOT the ‘incoming’
> ones for this chunk, but instead the ‘outgoing’ ones from the previous block ??
> Why??

Reviewer: ```suggestion
Yes, it is a little bit confusing, that receipts here are NOT the ‘incoming’
ones for this chunk, but instead the ‘outgoing’ ones from the previous block.  Why?!?!
```

> It all has to do with performance.

Reviewer: ```suggestion
This has to do with performance.
```

#### Simple approach

First, let’s imagine how the system would look like, if chunk contained things
that we’d expect:

* list of transactions
* list of incoming receipts
* list of outgoing receipts
* hash of the final state

This means, that the chunk producer has to compute all this information first,
before sending the chunk to other validators.

![image](https://user-images.githubusercontent.com/1711539/198282601-383977f1-08dd-45fe-aa19-70556d585034.png)

> Once other validators receive the chunk, they can start their own processing to

Reviewer: ```suggestion
Once the other validators receive the chunk, they can start their own processing to
```

verify those outgoing receipts/final state - and then do the signing. Only then,

> the next chunk producer can start creating the next chunk.

Reviewer: ```suggestion
can the next chunk producer start creating the next chunk.
```

While this approach does work, we can do it faster.

#### Faster approach

> What if the chunk didn’t contain the ‘output’ state? This changes a little bit
> our ‘mental’ model, as now when we’re singing the chunk, we’d be actually

Reviewer: ```suggestion
What if the chunk didn’t contain the ‘output’ state? This changes our ‘mental’ model
a little bit, as now when we’re singing the chunk, we’d actually be
```

verifying the previous chunk - but that’s the topic for the next article (TODO:
add future link to article about signatures and verification).

For now, imagine if the chunk only had:

* list of transactions
* list of incoming receipts

> In such case, the chunk producer could send the chunk a lot earlier, and

Reviewer: ```suggestion
In such a case, the chunk producer could send the chunk a lot earlier, and
```

> validators (and chunk producer) could do the processing at the same time:

Reviewer: ```suggestion
validators (and chunk producer) could do their processing at the same time:
```
